File: Habitica/src/main/java/com/habitrpg/android/habitica/models/user/Preferences.java
package com.habitrpg.android.habitica.models.user;


import com.habitrpg.android.habitica.models.AvatarPreferences;

import io.realm.RealmObject;
import io.realm.annotations.PrimaryKey;

public class Preferences extends RealmObject implements AvatarPreferences {

    @PrimaryKey
    private String userId;

    public Hair hair;
    SuppressedModals suppressModals;
    User user;
    private boolean costume, toolbarCollapsed, advancedCollapsed, tagsCollapsed, newTaskEdit, disableClasses, stickyHeader, sleep, hideHeader, dailyDueDefaultView;
    private String allocationMode, shirt, skin, size, background, chair, language, sound;
    private int dayStart, timezoneOffset;

    public Preferences() {
    }

    public String getBackground() {
        return background;
    }

    public void setBackground(String background) {
        this.background = background;
    }

    public int getDayStart() {
        return dayStart;
    }

    public void setDayStart(int dayStart) {
        this.dayStart = dayStart;
    }

    public boolean getCostume() {
        return costume;
    }

    public void setCostume(boolean costume) {
        this.costume = costume;
    }

    public boolean getToolbarCollapsed() {
        return toolbarCollapsed;
    }

    public void setToolbarCollapsed(boolean toolbarCollapsed) {
        this.toolbarCollapsed = toolbarCollapsed;
    }

    public boolean getAdvancedCollapsed() {
        return advancedCollapsed;
    }

    public void setAdvancedCollapsed(boolean advancedCollapsed) {
        this.advancedCollapsed = advancedCollapsed;
    }

    public boolean getTagsCollapsed() {
        return tagsCollapsed;
    }

    public void setTagsCollapsed(boolean tagsCollapsed) {
        this.tagsCollapsed = tagsCollapsed;
    }

    public boolean getNewTaskEdit() {
        return newTaskEdit;
    }

    public void setNewTaskEdit(boolean newTaskEdit) {
        this.newTaskEdit = newTaskEdit;
    }

    public boolean getDisableClasses() {
        return disableClasses;
    }

    public boolean getStickyHeader() {
        return stickyHeader;
    }

    public void setStickyHeader(boolean stickyHeader) {
        this.stickyHeader = stickyHeader;
    }

    public boolean getSleep() {
        return sleep;
    }

    public boolean getHideHeader() {
        return hideHeader;
    }

    public void setHideHeader(boolean hideHeader) {
        this.hideHeader = hideHeader;
    }

    public String getAllocationMode() {
        return allocationMode;
    }

    public void setAllocationMode(String allocationMode) {
        this.allocationMode = allocationMode;
    }

    public String getShirt() {
        return shirt;
    }

    public void setShirt(String shirt) {
        this.shirt = shirt;
    }

    public String getSkin() {
        return skin;
    }

    public void setSkin(String skin) {
        this.skin = skin;
    }

    public String getSize() {
        return size;
    }

    public void setSize(String size) {
        this.size = size;
    }

    public String getSound() {
        return sound;
    }

    public void setSound(String sound) {
        this.sound = sound;
    }

    public int getTimezoneOffset() {
        return timezoneOffset;
    }

    public void setTimezoneOffset(int timezoneOffset) {
        this.timezoneOffset = timezoneOffset;
    }

    public Hair getHair() {
        return hair;
    }

    public void setHair(Hair hair) {
        this.hair = hair;
    }

    public SuppressedModals getSuppressModals() {
        return suppressModals;
    }

    public void setSuppressModals(SuppressedModals suppressModals) {
        this.suppressModals = suppressModals;
    }

    public String getChair() {
        if (chair != null && !chair.equals("none")) {
            if (chair.length() > 5 && !chair.substring(0, 6).equals("chair_")) {
                return chair;
            } else {
                return "chair_" +chair;
            }
        }
        return null;
    }

    public void setChair(String chair) {
        this.chair = chair;
    }

    public boolean isSleep() {
        return sleep;
    }

    public void setSleep(boolean sleep) {
        this.sleep = sleep;
    }

    public boolean isDisableClasses() {
        return disableClasses;
    }

    public void setDisableClasses(boolean disableClasses) {
        this.disableClasses = disableClasses;
    }

    public String getLanguage() {
        return language;
    }

    public void setLanguage(String language) {
        this.language = language;
    }

    public boolean getDailyDueDefaultView() {
        return dailyDueDefaultView;
    }

    public void setDailyDueDefaultView(boolean dailyDueDefaultView) {
        this.dailyDueDefaultView = dailyDueDefaultView;
    }

    public String getUserId() {
        return userId;
    }

    public void setUserId(String userId) {
        this.userId = userId;
        if (hair != null && !hair.isManaged()) {
            hair.setUserId(userId);
        }
        if (suppressModals != null && !suppressModals.isManaged()) {
            suppressModals.setUserId(userId);
        }
    }
}

File: shared/src/commonMain/kotlin/com/habitrpg/shared/habitica/models/user/Preferences.kt
package com.habitrpg.shared.habitica.models.user


import com.habitrpg.shared.habitica.models.AvatarPreferences
import com.habitrpg.shared.habitica.nativePackages.NativeRealmObject
import com.habitrpg.shared.habitica.nativePackages.annotations.PrimaryKeyAnnotation
import com.habitrpg.shared.habitica.nativePackages.annotations.SerializedNameAnnotation

open class Preferences : NativeRealmObject(), AvatarPreferences {

    @PrimaryKeyAnnotation
    override var userId: String? = null
        set(userId: String?) {
            field = userId
            if (hair?.isManaged() == false) {
                hair?.userId = userId
            }
            if (suppressModals?.isManaged() == false) {
                suppressModals?.userId = userId
            }
        }

    override var hair: Hair? = null
    var suppressModals: SuppressedModals? = null
    override var costume: Boolean = false
    @SerializedNameAnnotation("disableClasses")
    override var disableClasses: Boolean = false
    @SerializedNameAnnotation("sleep")
    override var sleep: Boolean = false
    var dailyDueDefaultView: Boolean = false
    var automaticAllocation: Boolean = false
    var allocationMode: String? = null
    override var shirt: String? = null
    override var skin: String? = null
    override var size: String? = null
    override var background: String? = null
    override var chair: String? = null
    get() {
        return if (field != null && field != "none") {
            if (field?.contains("chair_") == true) {
                field
            } else {
                "chair_" + field!!
            }
        } else null
    }
    var language: String? = null
    var sound: String? = null
    var dayStart: Int = 0
    var timezoneOffset: Int = 0
    var timezoneOffsetAtLastCron: Int = 0
    var pushNotifications: PushNotificationsPreference? = null
    var emailNotifications: EmailNotificationsPreference? = null
    var autoEquip: Boolean = true


    fun hasTaskBasedAllocation(): Boolean = nativeHasTaskBasedAllocation(allocationMode, automaticAllocation)
}
expect fun nativeHasTaskBasedAllocation(allocationMode: String?, automaticAllocation: Boolean): Boolean
